src/bin/pg_autoctl/service_keeper.c

```diff
@@ -590,6 +499,144 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
+	bool reportPgIsRunning = ReportPgIsRunning(keeper);
+
+	/* We used to output that in INFO every 5s, which is too much chatter */
+	log_debug("Calling node_active for node %s/%d/%d with current state: "
+			  "%s, "
+			  "PostgreSQL %s running, "
+			  "sync_state is \"%s\", "
+			  "current lsn is \"%s\".",
+			  config->formation,
+			  keeperState->current_node_id,
+			  keeperState->current_group,
+			  NodeStateToString(keeperState->current_role),
+			  reportPgIsRunning ? "is" : "is not",
+			  postgres->pgsrSyncState,
+			  postgres->currentLSN);
+
+
+	/* ensure we use the correct retry policy with the monitor */
+	(void) pgsql_set_main_loop_retry_policy(&(monitor->pgsql));
+
+	/*
+	 * Report the current state to the monitor and get the assigned state.
+	 */
+	if (!monitor_node_active(monitor,
+							 config->formation,
+							 config->nodename,
+							 config->pgSetup.pgport,
+							 keeperState->current_node_id,
+							 keeperState->current_group,
+							 keeperState->current_role,
+							 reportPgIsRunning,
+							 postgres->currentLSN,
+							 postgres->pgsrSyncState,
+							 &assignedState))
+	{
+		log_error("Failed to get the goal state from the monitor");
+
+		/*
+		 * Check whether we're likely to be in a network partition.
+		 * That will cause the assigned_role to become demoted.
+		 */
+		if (keeperState->current_role == PRIMARY_STATE)
+		{
+			log_warn("Checking for network partitions...");
+
+			if (!is_network_healthy(keeper))
+			{
+				keeperState->assigned_role = DEMOTE_TIMEOUT_STATE;
+
+				log_info("Network in not healthy, switching to state %s",
+						 NodeStateToString(keeperState->assigned_role));
+			}
+			else
+			{
+				log_info("Network is healthy");
+			}
+		}
+
+		return false;
+	}
+
+	/*
+	 * We could contact the monitor, update our internal state.
+	 */
+	keeperState->last_monitor_contact = now;
+	keeperState->assigned_role = assignedState.state;
+
+	/* maybe update our cached list of other nodes */
+	if (!keeper_refresh_other_nodes(keeper, forceCacheInvalidation))
+	{
+		/*
+		 * We have a new MD5 but failed to update our list, try again next
+		 * round, the monitor might be restarting or something.
+		 */
+		log_error("Failed to update our list of other nodes");
+		return false;
+	}
+
+	/*
+	 * Also update the groupId and replication slot name in the
+	 * configuration file.
+	 */
+	(void) postgres_sprintf_replicationSlotName(assignedState.nodeId,
+												expectedSlotName,
+												sizeof(expectedSlotName));
+
+	if (assignedState.groupId != config->groupId ||
+		strneq(config->replication_slot_name, expectedSlotName))
+	{
+		bool postgresNotRunningIsOk = false;
+
+		if (!keeper_config_set_groupId_and_slot_name(config,
+													 assignedState.nodeId,
+													 assignedState.groupId))
+		{
+			log_error("Failed to update the configuration file "
+					  "with groupId %d and replication.slot \"%s\"",
+					  assignedState.groupId, expectedSlotName);
+			return false;
```

Comment:
Here's another behavioral change in the code caused by this refactoring. Before, when this failed we would error out right away and not do any state change. Now wo de not, but instead we set `couldContactMonitorThisRound` to false. 

It seems better not to error out, but try to do the state change instead. Setting `couldContactMonitorThisRound` to false, seems at least confusing. I think it makes more sense to return true in these cases, since the monitor could be contacted.

Reply:
I think it's better the way it is now. After thinking about it for a while, the main problem I see is the name of the boolean “CouldContactMonitor”... we kind of mean something else now. Will think about a better name, kept it that way for now.

Reply:
Can this cause any issues, where if calling `keeper_config_set_groupId_and_slot_name` fails we do not demote the primary when we should? Something, like what is happening now with `DEMOTE_TIMEOUT_STATE`.